### ISSUE TITLE
[PyTorch Upstream][PVC] 6 confirmed performance inductor drops ~ 15%

### ISSUE THREAD

**Issue by mengfei25** [intel/intel-xpu-backend-for-triton#8106]
## Summary
Triton-git00b9d1ae validation on PVC shows 6 confirmed drops compared with Triton-git1e2d42a0 test after applying rerun verification.
## Affected case
<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/kaileiyx/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/kaileiyx/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">

</head>

<body link="#467886" vlink="#96607D">

Suite | Dtype | Mode | Model | Ind Ratio | Eager Ratio
-- | -- | -- | -- | -- | --
timm_models | amp_bf16 | inference | gmixer_24_224 | 0.8597 | 0.9806
timm_models | amp_bf16 | inference | mixer_b16_224 | 0.8814 | 1.0028
timm_models | amp_fp16 | inference | gmixer_24_224 | 0.8542 | 1.0117
timm_models | float16 | training | resmlp_12_224 | 0.8483 | 1.0124
torchbench | amp_bf16 | inference | pyhpc_isoneutral_mixing | 0.8501 | 0.988
torchbench | amp_fp16 | inference | pyhpc_isoneutral_mixing | 0.8553 | 1.0028

</body>

</html>

## Reproducer
```bash
docker run -it -e TZ=Asia/Shanghai \
  --device=/dev/mem --device=/dev/dri --group-add video \
  --privileged --shm-size=8g \
  intelgpu/ubuntu-24.04-lts2:2523.40 bash

##Target triton
gh --repo pytorch/pytorch run download 34471485331 -p "triton-wheel-3.10-xpu-manylinux_2_28_x86_64"
 
curl -LsSf https://astral.sh/uv/install.sh | env UV_INSTALL_DIR="/usr/local/bin" sh
uv venv myvenv --python 3.10 --clear
source myvenv/bin/activate
uv pip install pip wheel "setuptools<81"

uv pip install torch==2.15.0.dev20260912+xpu torchvision==0.30.0.dev20260913+xpu torchaudio==2.11.0.dev20260913+xpu triton-xpu==3.8.0+git1e2d42a0 --pre --index-url https://download.pytorch.org/whl/nightly/xpu

pytorch_commit="$(python -c 'import torch; print(torch.version.git_version)')"
git clone https://github.com/pytorch/pytorch
cd pytorch
git checkout ${pytorch_commit}

uv pip install pandas psutil scipy requests
uv pip install -U numpy==1.26.4 accelerate
uv pip install -r .ci/docker/ci_commit_pins/huggingface-requirements.txt
uv pip install timm

git clone https://github.com/intel/torch-xpu-ops
rsync -avz ./torch-xpu-ops/.ci/benchmarks/ ./benchmarks/dynamo/
##benchmark
TORCHBENCH_COMMIT_ID="$(cat .ci/docker/ci_commit_pins/torchbench.txt 2> /dev/null || echo 'e03a63be')"
cd ..
git clone https://github.com/pytorch/benchmark benchmark
cd benchmark
python install.py --continue_on_fail pyhpc_isoneutral_mixing

pip install torchao==0.19.0.dev20260913+xpu --pre --index-url https://download.pytorch.org/whl/nightly/xpu
##Install target triton
uv pip uninstall triton-xpu -y
uv pip install /download/path/triton-xpu_*.wheel
##run command line
python benchmarks/dynamo/timm_models.py --performance -d xpu -n10 --backend=inductor --cold-start-latency --inference --amp --amp-dtype bfloat16 --only gmixer_24_224
python benchmarks/dynamo/timm_models.py --performance -d xpu -n10 --backend=inductor --cold-start-latency --inference --amp --amp-dtype bfloat16 --only mixer_b16_224
python benchmarks/dynamo/timm_models.py --performance -d xpu -n10 --backend=inductor --cold-start-latency --inference --amp --amp-dtype float16 --only gmixer_24_224
python benchmarks/dynamo/timm_models.py --performance -d xpu -n10 --backend=inductor --cold-start-latency --training --float16    --only resmlp_12_224
python benchmarks/dynamo/torchbench.py --performance -d xpu -n10 --backend=inductor --cold-start-latency --inference --amp --amp-dtype bfloat16 --only pyhpc_isoneutral_mixing
python benchmarks/dynamo/torchbench.py --performance -d xpu -n10 --backend=inductor --cold-start-latency --inference --amp --amp-dtype float16 --only pyhpc_isoneutral_mixing 
```

### Versions

torch 2.15.0.dev20260912+xpu
torchao 0.19.0.dev20260913+xpu
torchaudio 2.11.0.dev20260913+xpu
torchvision 0.30.0.dev20260913+xpu
triton-xpu 3.8.0+git00b9d1ae

**Comment by anmyachev** [comment]
Was able to reproduce on LTS for `gmixer_24_224`: 2.712x (3.8.0) vs 2.344x (3.9.0).

Command: `TORCHINDUCTOR_FORCE_DISABLE_CACHES=1 python benchmarks/dynamo/timm_models.py --performance -d xpu -n10 --backend=inductor --cold-start-latency --inference --amp --amp-dtype bfloat16 --only gmixer_24_224`

```bash
 git bisect good
3d59136d7c497098754e386d5c57c70d98ba25c3 is the first bad commit
commit 3d59136d7c497098754e386d5c57c70d98ba25c3
Author: Whitney Tsang <whitney.tsang@intel.com>
Date:   Wed Aug 19 18:12:38 2026 -0400

    Allow block_io on masked loads (#7205)
    
    PR #6752 added a guard preventing `block_io` on masked loads because the
    downstream lowering couldn't handle them. The actual issue was that
    de-anchoring masked loads triggered `RemoveLayoutConversions` to
    rematerialize them into a different encoding. This produced scalar
    (non-vectorized) loads through `PredicatedLoadINTEL`, which has a driver
    bug for strided accesses. The driver bug is now fixed.
```

@whitneywhtsang could you take a look?

Reverting this commit on top of release/3.9.0 returns perf to 2.693x (pretty close to what we have in 3.8.0 before)

**Comment by anmyachev** [comment]
For `pyhpc_isoneutral_mixing` model the regression is (`1.776x` vs `1.540x`):

```bash
git bisect bad
00b9d1aef6ec45fe29b5d16052268f02523f5ce0 is the first bad commit
commit 00b9d1aef6ec45fe29b5d16052268f02523f5ce0
Author: Ettore Tiotto <ettore.tiotto@intel.com>
Date:   Wed Sep 9 07:53:21 2026 -0400

    [XPU] Gate the auto-large-GRF retry on the reported n_spills (#7959)
    
    The auto-large-GRF rebuild recompiles a kernel at 256 GRF whenever the
    128-GRF build spills more than a threshold. That threshold was a flat 0
    bytes (#7615), so any spill at all triggered a recompile.
```

Command: `TORCHINDUCTOR_FORCE_DISABLE_CACHES=1 python benchmarks/dynamo/torchbench.py --performance -d xpu -n10 --backend=inductor --cold-start-latency --inference --amp --amp-dtype bfloat16 --only pyhpc_isoneutral_mixing`

@etiotto could you take a look?

**Comment by whitneywhtsang** [comment]
> Was able to reproduce on LTS for `gmixer_24_224`: 2.712x (3.8.0) vs 2.344x (3.9.0).
> 
> Command: `TORCHINDUCTOR_FORCE_DISABLE_CACHES=1 python benchmarks/dynamo/timm_models.py --performance -d xpu -n10 --backend=inductor --cold-start-latency --inference --amp --amp-dtype bfloat16 --only gmixer_24_224`
> 
>  git bisect good
> 3d59136d7c497098754e386d5c57c70d98ba25c3 is the first bad commit
> commit 3d59136d7c497098754e386d5c57c70d98ba25c3
> Author: Whitney Tsang <whitney.tsang@intel.com>
> Date:   Wed Aug 19 18:12:38 2026 -0400
> 
>     Allow block_io on masked loads (#7205)
>     
>     PR #6752 added a guard preventing `block_io` on masked loads because the
>     downstream lowering couldn't handle them. The actual issue was that
>     de-anchoring masked loads triggered `RemoveLayoutConversions` to
>     rematerialize them into a different encoding. This produced scalar
>     (non-vectorized) loads through `PredicatedLoadINTEL`, which has a driver
>     bug for strided accesses. The driver bug is now fixed.
> [@whitneywhtsang](https://github.com/whitneywhtsang) could you take a look?
> 
> Reverting this commit on top of release/3.9.0 returns perf to 2.693x (pretty close to what we have in 3.8.0 before)

It can be fixed by #8112.

**Comment by etiotto** [comment]
> For `pyhpc_isoneutral_mixing` model the regression is (`1.776x` vs `1.540x`):
> 
> git bisect bad
> 00b9d1aef6ec45fe29b5d16052268f02523f5ce0 is the first bad commit
> commit 00b9d1aef6ec45fe29b5d16052268f02523f5ce0
> Author: Ettore Tiotto <ettore.tiotto@intel.com>
> Date:   Wed Sep 9 07:53:21 2026 -0400
> 
>     [XPU] Gate the auto-large-GRF retry on the reported n_spills (#7959)
>     
>     The auto-large-GRF rebuild recompiles a kernel at 256 GRF whenever the
>     128-GRF build spills more than a threshold. That threshold was a flat 0
>     bytes (#7615), so any spill at all triggered a recompile.
> Command: `TORCHINDUCTOR_FORCE_DISABLE_CACHES=1 python benchmarks/dynamo/torchbench.py --performance -d xpu -n10 --backend=inductor --cold-start-latency --inference --amp --amp-dtype bfloat16 --only pyhpc_isoneutral_mixing`
> 
> [@etiotto](https://github.com/etiotto) could you take a look?

Fixed by PR #8115.  Cherry-pick for the release pls.

**Comment by anmyachev** [comment]
All fixes are in release/3.9.0 branch.
